Include/internal/pycore_pystate.h

```diff
@@ -300,6 +300,15 @@ PyAPI_FUNC(const PyConfig*) _Py_GetConfig(void);
 // See also PyInterpreterState_Get() and _PyInterpreterState_GET().
 extern PyInterpreterState* _PyGILState_GetInterpreterStateUnsafe(void);
 
+/* Modern equivalent of PyGILState_Check() for use in assertions only. */
+static inline int
+_Py_HoldsTstate(void)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    _Py_EnsureTstateNotNULL(tstate);
+    return 1;
+}
```

Comment:
```suggestion
static inline int
_Py_HoldsTstate(void)
{
    PyThreadState *tstate = _PyThreadState_GET();
    return tstate != NULL;
}
```

I don't think `_Py_EnsureTstateNotNULL` needs to be called as that would crash the interp if tstate doesn't exists. `PyGILState_Check` doesn't crash the interp in any case so this should behave similarly.

Reply:
I did this differently before. @vstinner prefers the nicer error message in case of a failure, but I like this approach too.